include/ogdf/basic/InducedSubgraph.h

```diff
@@ -0,0 +1,298 @@
+		newNodes++;
+		if (notifyObservers) {
+			m_regNodeArrays.keyAdded(v);
+			nodeInserted(cbData, vG, v);
+			for (GraphObserver* obs : getObservers()) {
+				obs->nodeAdded(v);
+			}
+		}
+	}
+}
+
+template<OGDF_NODE_ITER NI, OGDF_EDGE_ITER EI, bool copyEmbedding, bool copyIDs, bool notifyObservers>
+std::pair<int, int> Graph::insert(const NI& nodesBegin, const NI& nodesEnd, const EI& edgesBegin,
+		const EI& edgesEnd, NodeArray<node>& nodeMap, EdgeArray<edge>& edgeMap) {
+	OGDF_ASSERT(nodeMap.valid());
+	OGDF_ASSERT(edgeMap.valid());
+	OGDF_ASSERT(nodeMap.graphOf() == edgeMap.graphOf());
+	int newNodes = 0, newEdges = 0;
+	void* cbData = preInsert(copyEmbedding, copyIDs, notifyObservers, false, nodeMap, edgeMap,
+			&newNodes, &newEdges);
+	if (nodesBegin == nodesEnd) {
+		postInsert(cbData, newNodes, newEdges);
+		return {newNodes, newEdges};
+	}
+	insertNodes<NI, notifyObservers, copyIDs>(nodesBegin, nodesEnd, nodeMap, newNodes, cbData);
+
+	if (edgesBegin == edgesEnd) {
+		postInsert(cbData, newNodes, newEdges);
+		return {newNodes, newEdges};
+	}
+
+	if (!copyEmbedding && notifyObservers) {
+		int guessedEdges = internal::guess_dist(edgesBegin, edgesEnd);
+		if (guessedEdges > 0) {
+			m_regEdgeArrays.reserveSpace(guessedEdges);
+			m_regAdjArrays.reserveSpace(guessedEdges); // registry adds factor 2 in calculateArraySize
+		}
+	}
+
+	for (auto it = edgesBegin; it != edgesEnd; ++it) {
+		edge eG = *it;
+		node src = nodeMap[eG->source()];
+		node tgt = nodeMap[eG->target()];
+		if (src == nullptr || tgt == nullptr) {
+			continue;
+		}
+		if (copyIDs) {
+			m_edgeIdCount = max(m_edgeIdCount, eG->index() + 1);
+		}
+		// edgeMap[eG] is overwritten if it is != nullptr
+		edge e = edgeMap[eG] = pureNewEdge(src, tgt, copyIDs ? eG->index() : m_edgeIdCount++);
+		newEdges++;
+		if (!copyEmbedding) {
+			src->adjEntries.pushBack(e->m_adjSrc);
+			tgt->adjEntries.pushBack(e->m_adjTgt);
+			if (notifyObservers) {
+				m_regEdgeArrays.keyAdded(e);
+				m_regAdjArrays.keyAdded(e->adjSource());
+				edgeInserted(cbData, eG, e);
+				for (GraphObserver* obs : getObservers()) {
+					obs->edgeAdded(e);
+				}
+			}
+		}
+	}
+
+	if (!copyEmbedding) {
+#ifdef OGDF_HEAVY_DEBUG
+		consistencyCheck();
+#endif
+		postInsert(cbData, newNodes, newEdges);
+		return {newNodes, newEdges};
+	}
+
+	for (auto it = nodesBegin; it != nodesEnd; ++it) {
+		node vG = *it;
+		node v = nodeMap[vG];
+		for (adjEntry adjG : vG->adjEntries) {
+			edge eG = adjG->m_edge;
+			edge e = edgeMap[eG];
+			if (e == nullptr) {
+				continue;
+			}
+			adjEntry adj = adjG->isSource() ? e->adjSource() : e->adjTarget();
+			// edgeMap[eG] might be an old entry that was already inserted into the list
+			// so check whether adj is already in the list, indicated by having succ or pred,
+			// or being the only entry in the list
+			if (adj->succ() != nullptr || adj->pred() != nullptr || v->adjEntries.head() == adj) {
```

Comment:
Yes, the idea was to allow re-use of nodeMap and edgeMap, not only for multiple CCs or potentially even arbitrary subgraphs, but also for `GraphCopy.insert(G)` to directly pass its own maps instead of creating local ones on the stack. Unfortunately, this is not so easy for the filtering version, which is why the docs currently forbid this and `GraphCopy.insert` still uses additional maps on the stack. It would be nice if we could allow this in a well-defined manner, but that probably requires more work (esp. with different possible states in GraphCopy) and I'm for now not sure whether that is worth it for this niche use-case.